test/arbcall-test.jl

```diff
@@ -184,16 +198,27 @@ end
     z = Arb(0)
 
     Arblib.@arbcall_str "void arb_add(arb_t z, const arb_t x, const arb_t y, slong prec)"
-    @test typeof(Arblib.add!(z, x, y)) == Nothing
+    @test Arblib.add!(z, x, y) isa Nothing
+    @test Arblib.add!(
```

Comment:
I'm not sure this usage of `pointer_from_obj_ref` and calls to `Arblib.add!` correct? I think that julia metadata header in this call is interpreted as fields of `arb_struct`. maybe there weren't any pointers there, so you got lucky ;)  
```julia
julia> x = [1,2,3]
3-element Array{Int64,1}:
 1
 2
 3

julia> pl = Ptr{Clong}(pointer_from_objref(x))
Ptr{Int64} @0x00007f1f88ba95f0

julia> for i in 1:8
           @show unsafe_load(pl, i)
       end
unsafe_load(pl, i) = 139773414643248
unsafe_load(pl, i) = 3
unsafe_load(pl, i) = 559108
unsafe_load(pl, i) = 3
unsafe_load(pl, i) = 3
unsafe_load(pl, i) = 0
unsafe_load(pl, i) = 0
unsafe_load(pl, i) = 4294967296

julia> for i in 1:3
           @show unsafe_load(pl, sizeof(UInt)+i)
       end
unsafe_load(pl, sizeof(UInt) + i) = 1
unsafe_load(pl, sizeof(UInt) + i) = 2
unsafe_load(pl, sizeof(UInt) + i) = 3

```
whereas
```julia
julia> px = Base.unsafe_convert(Ptr{Clong}, Base.cconvert(Ptr{Clong}, x)) # exactly what `ccall` does
Ptr{Int64} @0x00007f1f88ba9630

julia> for i in 1:3
           @show unsafe_load(px, i)
       end
unsafe_load(px, i) = 1
unsafe_load(px, i) = 2
unsafe_load(px, i) = 3

```

this is why I said in https://github.com/kalmarek/Arblib.jl/issues/23#issuecomment-627566758 that the layout is not c-compatible and was thinking about implementing our own `unsafe_convert`. This wasn't necessary (there), because (?) `arb_struct` is immutable and stored inline:
```julia
julia> a = Arb("0.1")
[0.1000000000000000000000000000000000000000000000000000000000000000000000000000 +/- 1.95e-78]

julia> a.arb
Arblib.arb_struct(0xfffffffffffffffd, 0x0000000000000008, 0x0000000000000004, 0x0000000002d2fc60, 0xfffffffffffffefe, 0x0000000020000000)

julia> pa = Base.unsafe_convert(Ref{Arb}, Base.cconvert(Ref{Arb}, a))
Ptr{Arb} @0x00007fefaf5fae10

julia> pa_uint = Ptr{UInt}(pa)
Ptr{UInt64} @0x00007fefaf5fae10

julia> for i in 1:6
           @show unsafe_load(pa_uint, i)
       end
unsafe_load(pa_uint, i) = 0xfffffffffffffffd
unsafe_load(pa_uint, i) = 0x0000000000000008
unsafe_load(pa_uint, i) = 0x0000000000000004
unsafe_load(pa_uint, i) = 0x0000000002d2fc60
unsafe_load(pa_uint, i) = 0xfffffffffffffefe
unsafe_load(pa_uint, i) = 0x0000000020000000

julia> unsafe_load(pa_uint, 7) # our precision field
0x0000000000000100

julia> UInt(precision(a))
0x0000000000000100
```

Looking backwards: do we actually check the value of operations on `Arb`s anywhere in our tests?

Reply:
On the other hand :)
```julia
julia> pa = Base.unsafe_convert(Ref{Arb}, Base.cconvert(Ref{Arb}, a))
Ptr{Arb} @0x00007fefaf5fae10

julia> pointer_from_objref(a)
Ptr{Nothing} @0x00007fefaf5fae10

```
so this particular call returns the correct pointer? to be honest, this part of julia is still not really clear to me ;/ and I'm confused as to where did the header go?

Reply:
I had a lot of problems with this, it's very unclear to me how it works exactly. First I tried `pointer_from_objref(x)` but that didn't work, then I tried `pointer_from_objref(x.arb)` and that seemed to work so I went with that. 

But that it makes a difference if `arb_struct` being mutable or not makes sense. That explains some of the problems I encountered when preparing this change (I switched the from immutable to mutable now).  Maybe this change is suboptimal then? Since it changes the memory structure of `Arb` we might encounter problems later on?

Most of the tests I have written do not test the values of the `Arb`s but the test with the examples do test it (`test/examples.jl`). The main reason is that when I wrote most of the tests many of the required methods for checking the results was not implemented. I think this is something that will be fixed when we add a higher level interface.

Also doing some more testing it seems like there is a slight performance penalty to coming from this change :(

Reply:
the slowdown is probably due to the fact that `arb_struct` is no longer stored inline in `Arb`:
```julia
julia> a = Arb("0.1")
[0.1000000000000000000000000000000000000000000000000000000000000000000000000000 +/- 1.95e-78]

julia> cptr_a = Base.unsafe_convert(Ref{Arb}, Base.cconvert(Ref{Arb}, a))
Ptr{Arb} @0x00007f7c110a4370

julia> cptr_as = Base.unsafe_convert(Ref{Arblib.arb_struct}, Base.cconvert(Ref{Arblib.arb_struct}, a.arb))
Ptr{Arblib.arb_struct} @0x00007f7c12293750

julia> unsafe_load(Ptr{UInt}(cptr_a), 1)
0x00007f7c12293750

```